### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by Piyush4064**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    "@testing-library/jest-dom": "^5.16.5",
+    "@testing-library/react": "^13.4.0",
+    "@testing-library/user-event": "^13.5.0",
+    "react": "^18.2.0",
+    "react-dom": "^18.2.0",
+    "react-scripts": "5.0.1",
+    "web-vitals": "^2.1.4"
+  },
+  "scripts": {
+    "start": "react-scripts start",
```
Comment (nikesh04parekh) [inline package.json]: try running without react scripts command

File: src/App.css [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+.header {
+    width: 40%;
+    margin: 5% auto 0;
+}
+
+.header input {
+    font-size: 15px;
+    border: 2px solid #d1d3d4;
```
Comment (nikesh04parekh) [inline src/App.css]: use rem for responsiveness

File: data.txt [hunk 3]
```diff
@@ -0,0 +1,93 @@
+
+// const dummyTodos = [
+//     {
+//         id: 1,
```
Comment (Piyush4064) [inline data.txt]: Use category:[] this type of dataStructure.

File: package.json [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    "eject": "react-scripts eject"
+  },
+  "eslintConfig": {
+    "extends": [
+      "react-app",
+      "react-app/jest"
+    ]
+  },
+  "browserslist": {
+    "production": [
```
Comment (Piyush4064) [inline package.json]: Learn about all things in package.json.

File: src/App.js [hunk 5]
```diff
@@ -0,0 +1,179 @@
+import React, { useState } from "react";
+import Header from "./components/Header";
+import CardsForAllCategories from "./components/Cards/CardsForAllCategories";
+
+import "./App.css";
+
+const LOCAL_STORAGE_TODOS_KEY = "LOCAL_STORAGE_TODOS_KEY";
+const dummyTodos = JSON.parse(localStorage.getItem(LOCAL_STORAGE_TODOS_KEY)) || [];
+
+const LOCAL_STORAGE_TODO_CATEGORIES_KEY = "LOCAL_STORAGE_TODO_CATEGORIES_KEY";
```
Comment (Piyush4064) [inline src/App.js]: This should be in constant's file

File: src/App.js [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    const addNewCategory = (newCategory) => {
+        const clonedTodoCategories = addNewCategoryHelper(newCategory, [
+            ...todoCategories,
+        ]);
+        setTodoCategories(clonedTodoCategories);
+    };
+
+    const addNewTodo = (todoTask, category) => {
+        const clonedTodoItems = addNewTodoHelper(todoTask, category, { ...todoItems });
+        setTodoItems(clonedTodoItems);
```
Comment (nikesh04parekh) [inline src/App.js]: explore immer once.

File: src/App.js [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    handlePinButtonClickHelper,
+} from "./helperFunction";
+
+function App() {
+    const [todoItems, setTodoItems] = useState({});
+    const [todoCategories, setTodoCategories] = useState(["Work"]);
+
+    useEffect(() => {
+        const DUMMY_TODOS =
+            JSON.parse(localStorage.getItem(LOCAL_STORAGE_TODOS_KEY)) || {};
```
Comment (nikesh04parekh) [inline src/App.js]: Object.froze

File: src/components/Cards/CategoriesContainer.js [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+                        handleDeleteButtonClick={handleDeleteButtonClick}
+                        handlePinButtonClick={handlePinButtonClick}
+                    />
+                );
+            })}
+        </div>
+    );
+};
+
+export default CategoriesContainer;
```
Comment (nikesh04parekh) [inline src/components/Cards/CategoriesContainer.js]: memo

File: src/components/Cards/CategoriesContainer.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
+import React from "react";
```
Comment (nikesh04parekh) [inline src/components/Cards/CategoriesContainer.js]: Proptypes and defaultProps

File: src/components/Cards/CategoryCard.js [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+                        handleDeleteButtonClick={handleDeleteButtonClick}
+                        handlePinButtonClick={handlePinButtonClick}
+                    />
+                );
+            })}
+        </div>
+    );
+};
+
+export default CategoryCard;
```
Comment (nikesh04parekh) [inline src/components/Cards/CategoryCard.js]: memo

File: src/components/Cards/TodoCard.js [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    const handleDeleteButtonClick = () => {
+        props.handleDeleteButtonClick(props.id, props.category);
+    };
+    const handlePinButtonClick = () => {
+        props.handlePinButtonClick(props.id, props.category);
+    };
+
+    const handleInputSpanEnter = (event) => {
+        if (event.key === "Enter") {
+            props.handleInputSpanEnter(props.id, props.category, todoText);
```
Comment (nikesh04parekh) [inline src/components/Cards/TodoCard.js]: handleInputSpanEnter -> handleSaveTodo

File: src/components/Cards/TodoCard.js [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    const [showInputEle, setShowInputEle] = useState(false);
+
+    const handleCheckBoxClick = () => {
+        props.handleCheckBoxClick(props.id, props.category);
+    };
+
+    const handleDeleteButtonClick = () => {
+        props.handleDeleteButtonClick(props.id, props.category);
+    };
+    const handlePinButtonClick = () => {
```
Comment (nikesh04parekh) [inline src/components/Cards/TodoCard.js]: handlePinButtonClick -> handlePinTodo

File: src/components/Cards/TodoCard.js [hunk 13]
```diff
@@ -0,0 +1,64 @@
+import React, { useState } from "react";
+import ElementMaker from "./ElementMaker";
+import styles from "./todoCard.module.css";
+
+const TodoCard = (props) => {
+    const [todoText, setTodoText] = useState(props.todo);
+    const [showInputEle, setShowInputEle] = useState(false);
+
+    const handleCheckBoxClick = () => {
```
Comment (nikesh04parekh) [inline src/components/Cards/TodoCard.js]: handleCheckBoxClick -> handleCompleteTodo

File: src/components/Cards/TodoCard.js [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+
+    const handleDeleteButtonClick = () => {
+        props.handleDeleteButtonClick(props.id, props.category);
+    };
+    const handlePinButtonClick = () => {
+        props.handlePinButtonClick(props.id, props.category);
+    };
+
+    const handleInputSpanEnter = (event) => {
+        if (event.key === "Enter") {
```
Comment (nikesh04parekh) [inline src/components/Cards/TodoCard.js]: store EVENT_KEYS

File: src/components/Cards/TodoCard.js [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+        >
+            <input
+                type="checkbox"
+                className={styles.checkBox}
+                onClick={handleCheckBoxClick}
+                defaultChecked={props.isChecked}
+            />
+            <ElementMaker
+                value={todoText}
+                handleChange={(e) => setTodoText(e.target.value)}
```
Comment (nikesh04parekh) [inline src/components/Cards/TodoCard.js]: create a function in the functional component and use useCallback

File: src/components/Forms/AddTodoForm.js [hunk 16]
```diff
@@ -0,0 +1,42 @@
+import React, { useRef, useState } from "react";
+import CategorySelector from "./CategorySelector";
+import styles from "./addTodoForm.module.css";
+
+const AddTodoForm = ({ todoCategories, addNewTodo }) => {
+    const todoInput = useRef(null);
+    const [selectedTodoCategory, setSelectedTodoCategory] = useState("Work");
+
+    const handleCategoryChange = (event) => {
+        setSelectedTodoCategory(event.target.value);
```
Comment (nikesh04parekh) [inline src/components/Forms/AddTodoForm.js]: useCallback

File: src/helperFunction.js [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+    entry.id = new Date().valueOf();
+
+    entry.isPin
+        ? clonedTodoItems[category].unshift(entry)
+        : clonedTodoItems[category].push(entry);
+    saveTodoItems(clonedTodoItems);
+    return clonedTodoItems;
+};
+
+export const createTodo = (todo) => {
```
Comment (nikesh04parekh) [inline src/helperFunction.js]: builder

File: src/helperFunction.js [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+    );
+}
+
+const findTodoIndex = (todoId, category, todoItems) => {
+    for (let todoIndex = 0; todoIndex < todoItems[category].length; todoIndex++) {
+        if (todoItems[category][todoIndex].id === todoId) {
+            return todoIndex;
+        }
+    }
+    return -1;
```
Comment (nikesh04parekh) [inline src/helperFunction.js]: forEach function from Array Prototype

File: src/helperFunction.js [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+
+export function capitalizeFirstLetter(str) {
+    return str.charAt(0).toUpperCase() + str.slice(1).toLowerCase();
+}
+
+function saveTodoItems(todoItems) {
+    localStorage.setItem(LOCAL_STORAGE_TODOS_KEY, JSON.stringify(todoItems));
+}
+
+export function saveTodoCategories(todoCategories) {
```
Comment (nikesh04parekh) [inline src/helperFunction.js]: saveTodoCategories -> updateDataInStore

File: src/App.css [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+.fa-solid,
+.fa-trash,
+.fa-solid,
+.fa-thumbtack {
+    pointer-events: none;
+}
+
+.todoCardExpand {
```
Comment (nikesh04parekh) [inline src/App.css]: not used

File: src/App.css [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+.header input {
+    font-size: 1rem;
+    border: 0.1rem solid #d1d3d4;
+    padding: 0.7rem;
+    color: #111111;
+    font-weight: 500;
+    border-radius: 0.3rem;
+}
+
+.fa-solid,
```
Comment (nikesh04parekh) [inline src/App.css]: there is a selector to apply styles based on prefix. search that

File: src/App.js [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+import { STORE } from "./constant";
+
+import {
+    addNewCategory,
+    addNewTodo,
+    completeTodo,
+    saveTodo,
+    deleteTodo,
+    pinTodo,
+} from "./helperFunction";
```
Comment (nikesh04parekh) [inline src/App.js]: rename

File: src/App.js [hunk 23]
```diff
@@ -0,0 +1,105 @@
...
+        const DUMMY_TODOS =
+            JSON.parse(localStorage.getItem(STORE.LOCAL_STORAGE_TODOS_KEY)) || {};
+        const DUMMY_TODO_CATEGORIES = JSON.parse(
+            localStorage.getItem(STORE.LOCAL_STORAGE_TODO_CATEGORIES_KEY)
+        ) || ["Work"];
+        setTodoItems(DUMMY_TODOS);
+        setTodoCategories(DUMMY_TODO_CATEGORIES);
+    }, []);
+
+    const updateTodoItems = (clonedTodoItems) => {
```
Comment (nikesh04parekh) [inline src/App.js]: check this why we need to create wrapper function

File: src/components/cards/category/Category.js [hunk 24]
```diff
@@ -0,0 +1,69 @@
...
+    updateTodoItems,
+    category,
+    handleCompleteTodo,
+    handleSaveTodo,
+    handleDeleteTodo,
+    handlePinTodo,
+}) => {
+    function handleOnDrop(event) {
+        const receivedTodo = JSON.parse(event.dataTransfer.getData("transferTodo"));
+        if (receivedTodo.category === category) {
```
Comment (nikesh04parekh) [inline src/components/cards/category/Category.js]: move to helpers and segregrate the functionalities

File: src/components/cards/category/category.module.css [hunk 25]
```diff
@@ -0,0 +1,11 @@
+.card {
+    text-align: center;
+    width: 31%;
+    min-width: 300px;
```
Comment (nikesh04parekh) [inline src/components/cards/category/category.module.css]: rem

File: src/components/cards/category/constant.js [hunk 26]
```diff
@@ -0,0 +1,8 @@
+const STORE = {
```
Comment (Piyush4064) [inline src/components/cards/category/constant.js]: use freeaze here directly

File: src/components/cards/category/helperFunction.js [hunk 27]
```diff
@@ -0,0 +1,125 @@
+import { STORE } from "./constant";
+import { produce } from "immer";
+
+function updateTodoItemsInStore(todoItems) {
+    localStorage.setItem(STORE.LOCAL_STORAGE_TODOS_KEY, JSON.stringify(todoItems));
+}
+
+function findTodoIndex(todoId, category, todoItems) {
+    let index = -1;
+    todoItems[category].forEach((todoItem, idx) => {
```
Comment (nikesh04parekh) [inline src/components/cards/category/helperFunction.js]: use inbuilt findindex

File: src/components/cards/category/helperFunction.js [hunk 28]
```diff
@@ -0,0 +1,125 @@
...
+}
+
+export function addNewTodo(todoTask, category, todoItems) {
+    if (todoTask.trim() === "") {
+        return todoItems;
+    }
+    const isPresent = todoItems.hasOwnProperty(category);
+    const newTodo = new createTodo(todoTask);
+    let clonedTodoItems = produce(todoItems, (draft) => {
+        if (!isPresent) {
```
Comment (nikesh04parekh) [inline src/components/cards/category/helperFunction.js]: `const prevTodoOfCategory = draft[category] || [];
draft[category] = [...prevTodoOfCategory, newTodo.toJSON()]`

File: src/components/cards/category/helperFunction.js [hunk 29]
```diff
@@ -0,0 +1,125 @@
...
+        }
+    });
+    return index;
+}
+
+export function addNewTodo(todoTask, category, todoItems) {
+    if (todoTask.trim() === "") {
+        return todoItems;
+    }
+    const isPresent = todoItems.hasOwnProperty(category);
```
Comment (nikesh04parekh) [inline src/components/cards/category/helperFunction.js]: not needed

File: src/components/cards/category/helperFunction.js [hunk 30]
```diff
@@ -0,0 +1,125 @@
...
+    const todoIndex = findTodoIndex(todoId, category, todoItems);
+    if (todoIndex === -1) {
+        return todoItems;
+    }
+
+    const clonedTodoItems = produce(todoItems, (draft) => {
+        const entry = draft[category][todoIndex];
+        draft[category].splice(todoIndex, 1);
+        entry.isPin = !entry.isPin;
+        entry.id = new Date().valueOf();
```
Comment (nikesh04parekh) [inline src/components/cards/category/helperFunction.js]: not needed

File: src/components/cards/category/helperFunction.js [hunk 31]
```diff
@@ -0,0 +1,125 @@
...
+        draft[category].splice(todoIndex, 1);
+        if (draft[category].length === 0) {
+            delete draft[category];
+        }
+    });
+    updateTodoItemsInStore(clonedTodoItems);
+    return clonedTodoItems;
+}
+
+export class createTodo {
```
Comment (nikesh04parekh) [inline src/components/cards/category/helperFunction.js]: builder

File: src/components/cards/todo/Todo.js [hunk 32]
```diff
@@ -0,0 +1,92 @@
+import React, { useState, useCallback } from "react";
```
Comment (nikesh04parekh) [inline src/components/cards/todo/Todo.js:1]: import order not correct

File: src/components/forms/categorySelector/CategorySelector.js [hunk 33]
```diff
@@ -0,0 +1,24 @@
+import React from "react";
+import styles from "./categorySelector.module.css";
+
+const CategorySelector = ({ value, handleCategoryChange, todoCategories }) => {
+    return (
+        <>
```
Comment (nikesh04parekh) [inline src/components/forms/categorySelector/CategorySelector.js]: can be removed

**Review by nikesh04parekh — APPROVED** [review]

**Pull request merged**